conda/recipes/libcuml/meta.yaml

```diff
@@ -34,12 +34,14 @@ requirements:
     - cudf {{ minor_version }}
     - cudatoolkit {{ cuda_version }}.*
     - libcumlprims {{ minor_version }}
+    - ucx-py {{ minor_version }}
     - lapack
     - protobuf >=3.4.1,<4.0.0
   run:
     - libcumlprims {{ minor_version }}
     - cudf {{ minor_version }}
     - nccl 2.4.*
+    - ucx-py {{ minor_version }}
```

Comment:
Wouldn't still requiring the `ucx-py` package in the conda recipe make it so that the `libnuma` problem still happen? i.e. the `dlopen` approach makes it so that `libcumlcomms` will  check for libucp.so which is the guy that actually linked to libnuma and caused the issue, so the libnuma issue could still happen, no? I think the full solution would be to also check for libnuma in ucp_helper with dlopen at the same time to be able to gracefully err if it is libnuma the thing that is missing until/if libucp uses dlopen itself to do that check, what do you think?

Reply:
It's very possible that my understanding of where and when the `libnuma` occurs is not correct. I got the impression that it was the `rpath` being explicitly linked in our cython build, through the use of `libs.append("libucp.so")`, that was causing the `libnuma` error. I thought our original plan was to prevent the error by disabling ucx altogether in this case, by not installing the conda package. Is the error actually happening when the `ucx` / `ucx-py` conda package is installed?

cc @rlratzel for more details. 

With the changes in the PR, any runtime error that results from loading the required ucp symbols will be caught in an `ASSERT` and the Python layer *should* continue business as usual. I'll also provide a test to be sure the failure is, in fact, graceful.

If it does fail gracefully, the assertion error message will propagate to the users, which will let them know libnuma wasn't installed. We could make a prettier error message to let them know what happened, or I can add some documentation in Github around the error. My only hesitation with doing an explicit check for each and every possible failure case is that the list of failure conditionals could get unwieldy quite quickly.

Reply:
I believe your understanding is correct regarding the `rpath` situation in cython indeed, which was an issue for the (to my knowlede somewhat rare) situation where libnuma is not present in the system, and the plan you mention indeed is correct. But adding the `ucx-py` requirement here makes it so that the conda `ucx-py` package will always be installed, and only checking for `libucp` in `dlopen` (while `ucx` doesn't have such check yet) could end up leading to the same situation where `libnuma` is not installed (but `libucp` is) and would/could cause an issue (maybe I'm wrong in that assumption?). 

Since `libnuma` is the issue that originated all of this, my proposal is that we draw the line here, where this is the only additional check  beyond `libucp`, due to being the known issue, and therefore disable ucx functionality if there is no `libnuma` for the time being just to be sure that we don't run into the same situation as before just in a different step.

I think @rlratzel is away this week but his input would be nice...

Reply:
> But adding the ucx-py requirement here makes it so that the conda ucx-py package will always be installed, and only checking for libucp in dlopen (while ucx doesn't have such check yet) could end up leading to the same situation where libnuma is not installed (but libucp is) and would/could cause an issue (maybe I'm wrong in that assumption?).

From what I recall, the meeting we had w/ ops presented 2 options- one option was to provide 2 different configurations where the conda packages could be installed with and without ucx. The other option was to load the needed symbols from the underlying ucp package lazily. Since we know the use of UCX is going to continue to increase across the RAPIDS projects, our choice of the second option implied, to me, that we weren't going to be changing the conda dependencies.

Normally, when a cython library is imported, all the needed symbols for the tree of imports are loaded from the underlying C++ libraries eagerly. This is why we often get the symbol errors at import time, rather than when they are used. From the `libnuma` discussion with @rlratzel and @jakirkham during release 0.11, it seemed like the UCX error was occurring when the comms symbols were being imported (and subsequently loaded from `libcuml`). The integration of `dlopen` in this PR causes those symbols to be looked up lazily, only when the `isend` / `irecv` calls are actually made on the communicator.

We have encapsulated all of our interactions with `ucx-py` in `cuml/dask/common/comms.py` and catch the error when ucp cannot be imported successfully in `comms.py`:

```
if is_ucx_enabled() and has_ucp():
    import ucp
```

What do you think?

It will be nice if the `ucx` project could use `dlopen` to lazily load the `libnuma` symbols where they are invoked in the code, and perhaps use the slower mechanism when `libnuma` is not installed. I suppose they might even be able to do this without `dlopen`.

Reply:
I'm not an expert on the exact implementation details, but what @cjnolet has described sounds correct. cuML should be able to have a hard dependency on `ucx-py`, but use dlopen to access `libucp` so `libcuml` isn't directly linked to it.

So I think adding the conda dependency for ucx-py should be fine.

Reply:
It took me a second to follow this, but after reading commit ( https://github.com/rapidsai/cuml/commit/6a30f6d89d4ec34b59a31ec834b0d92d8f594c33 ) I think this makes sense. Basically you are checking if the `libucp.so` can be loaded in C and Python uses that same info to determine if it can use it as well. Looks good. Thanks Corey! 😄